### PR TITLE
github: refactor API client to only associate with one token

### PR DESCRIPTION
**Pull request by unknwon**
This is pure refactoring and should not change any functionality, it simplifies the design of the GitHub client to only associate with exactly one token.

## Why?

1. With the introduction of `WithToken` method, it is not necessary to pass around the `token` argument anymore.
2. Simplifies the management of repository cache, instead of holding a map of caches for different tokens, just one cache object that is associated with the token. Much easier to reason.

---

~~**NOTE**: I'll fix failing tests that require a bit more effort once get approval for the core change.~~

### CODE REVIEW

File: internal/extsvc/github/client.go [hunk 1]
```diff
@@ -154,10 +143,10 @@ func NewClient(apiURL *url.URL, defaultToken string, cli httpcli.Doer) *Client {
 	return &Client{
 		apiURL:       apiURL,
 		githubDotCom: urlIsGitHubDotCom(apiURL),
-		defaultToken: defaultToken,
+		token:        token,
 		httpClient:   cli,
 		RateLimit:    &ratelimit.Monitor{HeaderPrefix: "X-"},
-		repoCache:    map[string]*rcache.Cache{},
+		repoCache:    NewRepoCache(apiURL, token, "", 0),
```
Comment (ryanslade) [inline internal/extsvc/github/client.go:149]: What effect will having a TTL of zero have?

Comment (unknwon) [reply internal/extsvc/github/client.go:149]: I think it means never expire.
https://github.com/sourcegraph/sourcegraph/blob/7dc1b314b3084a6ab508d427f7631f87916239ce/internal/rcache/rcache.go#L103

File: internal/extsvc/github/client.go [hunk 2]
```diff
@@ -55,25 +54,15 @@ type Client struct {
...
-	repoCacheMu sync.RWMutex
-
-	// repoCachePrefix is the cache key prefix used when constructing a new rcache.Cache instance
-	// for repoCache. It should normally be left empty. It is only used in tests. Specifying a
-	// non-empty value will result in different Redis key values.
-	repoCachePrefix string
-
-	// repoCacheTTL is the TTL of cache entries.
-	repoCacheTTL time.Duration
+	// repoCache is the repository cache associated with the token.
```
Comment (beyang) [inline internal/extsvc/github/client.go:64]: Is it possible for the token to ever change? If so, then is there a cache invalidation mechanism? If not, could we add a docstring to the struct indicating it should not be modified after creation?

File: internal/extsvc/github/client.go (hunk 1 above)
Comment (beyang) [reply internal/extsvc/github/client.go:149]: Could you document that TTL=0 means no expiration in the docstring for `NewRepoCache`?

Comment (unknwon) [reply internal/extsvc/github/client.go:149]: I'm wondering if we can get rid of cache key prefix and cache TTL (last two arguments, i.e. `"", 0`) from the `NewRepoCache` signature entirely. They are only used in tests, and as we always sets up a clean state of rcache (`rcache.SetupForTest` for each test that requires it), I don't see a need to have the prefix and TTL. WDYT?

File: internal/extsvc/github/client.go (hunk 2 above)
Comment (unknwon) [reply internal/extsvc/github/client.go:64]: Token should never change with the same client, i.e. always create a new client with new `repoCache` for a different token. I'll document it.

File: internal/extsvc/github/client.go (hunk 1 above)
Comment (unknwon) [reply internal/extsvc/github/client.go:149]: Sorry I was wrong, 0 actually means default depending on whether GitHub.com or not:
https://github.com/sourcegraph/sourcegraph/blob/7dc1b314b3084a6ab508d427f7631f87916239ce/internal/extsvc/github/client.go#L108-L115

**Comment by unknwon** [comment]
@beyang I would appreciated another look on the `_test.go` files. 

Main change: I removed the `mockGitHub` because we're able to mock the GitHub client directly, so instead of testing our mock logic (zero relation to the actual code being tested), we return desired results (i.e. `repos map[string]*github.Repository`) directly.

**Comment by unknwon** [comment]
Merging as-is since for authz, only test code is changed (i.e. no change to the code logic, just how we test it) so it is low risk to address comments in post-merge.

**Pull request merged**